File: runtimes/node/src/rabbitmq.ts
import * as amqp from 'amqplib';

export class RabbitMQAsyncSink {
  private connection: amqp.ChannelModel | null = null;
  private channel: amqp.Channel | null = null;

  constructor(private readonly url: string) {}

  async connect(): Promise<void> {
    if (!this.connection) {
      this.connection = await amqp.connect(this.url);
      this.channel = await this.connection.createChannel();
    }
  }

  async close(): Promise<void> {
    if (this.channel) {
      await this.channel.close();
      this.channel = null;
    }
    if (this.connection) {
      await this.connection.close();
      this.connection = null;
    }
  }

  async publishAsync(queue: string, message: any): Promise<void> {
    if (!this.channel) {
      throw new Error('Not connected. Call connect() first.');
    }
    await this.channel.assertQueue(queue, { durable: true });
    this.channel.sendToQueue(queue, Buffer.from(JSON.stringify(message)));
  }
}

File: runtimes/python/helix_rt/rabbitmq.py
import aio_pika
import json

class RabbitMQAsyncSink:
    def __init__(self, connection_url: str, routing_key: str):
        self.connection_url = connection_url
        self.routing_key = routing_key
        self.connection = None
        self.channel = None

    async def connect(self):
        if not self.connection:
            self.connection = await aio_pika.connect_robust(self.connection_url)
            self.channel = await self.connection.channel()

    async def publish_async(self, message: dict):
        await self.connect()
        await self.channel.default_exchange.publish(
            aio_pika.Message(body=json.dumps(message).encode('utf-8')),
            routing_key=self.routing_key,
        )

    async def close(self):
        if self.connection:
            await self.connection.close()
